spring-rabbit/src/main/java/org/springframework/amqp/rabbit/support/PublisherCallbackChannelImpl.java

```diff
@@ -691,32 +687,33 @@ private synchronized void generateNacksForPendingAcks(String cause) {
 				logger.debug("Added listener " + listener);
 			}
 		}
-		return this.pendingConfirms.get(listener);
 	}
 
 	@Override
-	public synchronized boolean removeListener(Listener listener) {
-		Listener mappedListener = this.listeners.remove(listener.getUUID());
-		boolean result = mappedListener != null;
-		if (result && this.listeners.size() == 0) {
-			this.delegate.removeConfirmListener(this);
-			this.delegate.removeReturnListener(this);
+	public synchronized Collection<PendingConfirm> expire(Listener listener, long cutoffTime) {
+		SortedMap<Long, PendingConfirm> pendingConfirmsForListener = this.pendingConfirms.get(listener);
+		if (pendingConfirmsForListener == null) {
+			return Collections.<PendingConfirm>emptyList();
 		}
-		Iterator<Entry<Long, Listener>> iterator = this.listenerForSeq.entrySet().iterator();
-		while (iterator.hasNext()) {
-			Entry<Long, Listener> entry = iterator.next();
-			if (entry.getValue() == listener) {
-				iterator.remove();
+		else {
+			synchronized (pendingConfirmsForListener) {
```

Comment:
We're not synchronized on the local variable, we are synchronized on the `ConcurrentHashmap` it holds a reference to. 

However, you are correct; I was thinking of `SynchronizedMap` where you have to synchronize while iterating.

It is not necessary to synchronize while iterating over a `ConcurrentHashMap`.